pkg/controller/nodemaintenance/master_quorum_check.go

```diff
@@ -0,0 +1,105 @@
+package nodemaintenance
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	policy "k8s.io/api/policy/v1beta1"
+	kubernetes "k8s.io/client-go/kubernetes"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+const OpenshiftMachineConfigNamespace = "openshift-machine-config-operator"
+
+func checkValidQuorum(client kubernetes.Interface, node *corev1.Node) (bool, bool, error) {
+	if isMasterNode(client, node) {
+		validQuorum, fallbackPDB, err := checkValidQuorumPdb(client)
+		if err != nil {
+			return false, false, err
+		}
+		log.Infof("node %s is a master. quorum status: %t", node.Name, validQuorum)
+		return validQuorum, fallbackPDB, nil
+	} else {
+		log.Infof("node %s is not a master", node.Name)
+	}
+	return true, false, nil
+}
+
+func findMcoPDBForEtcd(client kubernetes.Interface) (*policy.PodDisruptionBudget,bool,error) {
+	pdbList, err := client.Policy().PodDisruptionBudgets(OpenshiftMachineConfigNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, false, fmt.Errorf("can't get mco pdb: %v", err)
+	}
+	if pdbList.Size() == 0  || pdbList.Items == nil || len(pdbList.Items) == 0 {
+		return nil, true, fmt.Errorf("mco pdb is empty")
+	}
+	return &pdbList.Items[0], false, nil
+}
+
+func isMasterNode(client kubernetes.Interface, node *corev1.Node) bool {
+	if node.ObjectMeta.Labels != nil {
+		_, isMaster := node.ObjectMeta.Labels[MasterNodeLabel]
+		return isMaster
+	}
+	return false
+}
+
+func createFallbackPDB(client kubernetes.Interface) (*policy.PodDisruptionBudget,error) {
+
+	objectName  := "etcpdb"
+	minAvailable := intstr.FromInt(1)
+	object := &policy.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objectName,
+			Namespace: NodeMaintenanceOperatorNamespace,
+		},
+		Spec: policy.PodDisruptionBudgetSpec{
+			Selector:     &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"k8s-app": "etcd-quorum-guard",
+				},
+			},
+			MinAvailable: &minAvailable,
+		},
+		Status: policy.PodDisruptionBudgetStatus{},
+	}
+
+	_, err := client.PolicyV1beta1().PodDisruptionBudgets(NodeMaintenanceOperatorNamespace).Create(object)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			object, err = client.PolicyV1beta1().PodDisruptionBudgets(NodeMaintenanceOperatorNamespace).Get(objectName,metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return object, nil
+}
+
+func checkValidQuorumPdb(client kubernetes.Interface) (bool, bool, error) {
+
+	fallbackPDB := false
+	pdb, pdbNotFound, err := findMcoPDBForEtcd(client)
+	if pdbNotFound {
+		fallbackPDB = true
+		pdb, err = createFallbackPDB(client)
+	}
```

Comment:
Remember when I mentioned creating this as part of the test harness, not in Go code?

Reply:
done

Reply:
done, also moved the this e2e test case  into a separate file.

Reply:
I'm confused. You say done, but the logic that creates a PDB is still an active part of the flow.

Reply:
I misunderstoood that  we needed a fallback option if the operator is running on an old cluster (k8s without mco) or old cluster (os 3.11). Apparently that is incorrect, so I will delete.

Reply:
so if the node is a master node and we can't get the pdb of the mco, then remediation failed. Is that correct?